File: Go/quicksort.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"runtime"
	"strconv"
	"time"
	"sync"
)

// Função para obter o uso de memória
func getMemoryUsage() int64 {
	var memStats runtime.MemStats
	runtime.ReadMemStats(&memStats)
	return int64(memStats.Sys / 1024) // Retorna em KB
}

// Função de partição de Hoare
func partition(arr []int, low, high int) int {
	pivot := arr[low]
	i := low - 1
	j := high + 1
	for {
		i++
		for arr[i] < pivot {
			i++
		}
		j--
		for arr[j] > pivot {
			j--
		}
		if i >= j {
			return j
		}
		arr[i], arr[j] = arr[j], arr[i]
	}
}

// Função de QuickSort paralelo
func quickSortParallel(arr []int, low, high int, wg *sync.WaitGroup) {
	defer wg.Done()

	if low < high {
		pi := partition(arr, low, high)

		// Dividir a tarefa entre duas Goroutines
		wg.Add(2) // Adicionar duas Goroutines para espera
		go quickSortParallel(arr, low, pi, wg)
		go quickSortParallel(arr, pi+1, high, wg)
	}
}

func main() {
	// Medir o tempo de execução
	start := time.Now()

	// Abrir o arquivo de entrada
	file, err := os.Open("arq.txt")
	if err != nil {
		fmt.Println("Erro ao abrir o arquivo:", err)
		return
	}
	defer file.Close()

	// Ler os números do arquivo
	var arr []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		num, err := strconv.Atoi(scanner.Text())
		if err != nil {
			fmt.Println("Erro ao ler o número:", err)
			return
		}
		arr = append(arr, num)
	}

	// Usar um WaitGroup para gerenciar as Goroutines
	var wg sync.WaitGroup
	wg.Add(1) // Inicializa o WaitGroup com 1 Goroutine principal

	// Executar o QuickSort paralelo
	quickSortParallel(arr, 0, len(arr)-1, &wg)

	// Esperar todas as Goroutines terminarem
	wg.Wait()

	// Salvar o resultado no arquivo de saída
	outputFile, err := os.Create("arq-saida.txt")
	if err != nil {
		fmt.Println("Erro ao abrir o arquivo de saída:", err)
		return
	}
	defer outputFile.Close()

	for _, num := range arr {
		_, err := fmt.Fprintln(outputFile, num)
		if err != nil {
			fmt.Println("Erro ao escrever no arquivo de saída:", err)
			return
		}
	}

	// Medir o tempo de execução
	duration := time.Since(start)
	fmt.Printf("Tempo de execução: %v segundos\n", duration.Seconds())

	// Obter o uso de memória
	finalMemory := getMemoryUsage()
	fmt.Printf("Uso de memória final: %d KB\n", finalMemory)
}

File: C/atv3/quicksort.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
#include <ctime>
#include <fstream>
#include <chrono>
#include <sys/resource.h>

using namespace std;

long getMemoryUsage() {
    struct rusage usage;
    getrusage(RUSAGE_SELF, &usage);
    return usage.ru_maxrss;
}

// Função para trocar dois elementos
void swap(int &a, int &b) {
    int temp = a;
    a = b;
    b = temp;
}

// Função de partição de Hoare
int partition(vector<int>& arr, int low, int high) {
    int pivot = arr[low];
    int i = low - 1;
    int j = high + 1;
    while (true) {
        do { i++; } while (arr[i] < pivot);
        do { j--; } while (arr[j] > pivot);
        if (i >= j) return j;
        swap(arr[i], arr[j]);
    }
}

// Função do QuickSort sequencial
void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);

        // Recursão nas duas metades
        quickSort(arr, low, pi);
        quickSort(arr, pi + 1, high);
    }
}

int main() {
    auto start = chrono::high_resolution_clock::now();

    ifstream file("arq.txt");
    if (!file) {
        cerr << "Erro ao abrir o arquivo!" << endl;
        return 1;
    }

    vector<int> arr;
    int num;
    while (file >> num) {
        arr.push_back(num);
    }
    file.close();

    srand(time(NULL));  // Garantir que a semente seja diferente a cada execução

    // Chamar o QuickSort sequencial
    quickSort(arr, 0, arr.size() - 1);

    ofstream outputFile("arq-saida.txt");
    if (!outputFile) {
        cerr << "Erro ao abrir o arquivo de saída!" << endl;
        return 1;
    }

    // Escrever o array ordenado no arquivo de saída
    for (int i = 0; i < arr.size(); i++) {
        outputFile << arr[i] << endl;
    }
    outputFile.close();

    auto end = chrono::high_resolution_clock::now();
    chrono::duration<double> duration = end - start;
    cout << "Tempo de execução: " << duration.count() << " segundos" << endl;

    long finalMemory = getMemoryUsage();
    cout << "Uso de memória final: " << finalMemory << " KB" << endl;

    return 0;
}
